### ISSUE TITLE
catkin do not install python packages

### ISSUE THREAD

**Issue by severin-lemaignan** [ros/catkin#500]
According to [catkin documentation](http://ros.org/doc/groovy/api/catkin/html/user_guide/setup_dot_py.html) (and real testing :-) ), python modules declared in a `setup.py` are not installed. Instead a relay package is created, that import the original Python package _from the source prefix_. Unless I missing something, this behaviour prevent installation of Python packages in the install prefix, and thus prevent proper packaging/distribution/cross-compilation/etc.

**Comment by wjwwood** [comment]
This is only done in the `devel` space. When you invoke the install target the python package is installed correctly.

**Comment by wjwwood** [comment]
The `devel` space is designed to work like this so you can have the contents of the `devel` folder "look like" an installation of your package, but with references to your source space. The benefit of this is that you can use the devel space like it was installed, but you can iterate on Python more quickly by changing the code in the source space.

Once installed, the code has no reference to the source, build, or devel spaces.

**Comment by severin-lemaignan** [comment]
Sorry, I've not been accurate enough in my issue description.

Here are the steps I've followed:

```
$ cd <metapkg>/<pkg> && mkdir build
$ cd build
$ cmake -DCMAKE_INSTALL_PREFIX=/tmp/test ..
$ make install 
```

Doing so (which looks for me like 'installing') generate a relay Python package that points to my source directory.
I can reliably reproduce the error.

More troubling, however, is that the faulty behaviour only occurs on one (virtual) machine, not on my main dev machine. Both use the same Groovy version (with catkin 0.5.71).

Do you have any hints that would help me understanding what can occur here?

**Comment by dirk-thomas** [comment]
Could you please post a link to the sources of your package?

**Comment by wjwwood** [comment]
I cannot reproduce this with this package:

https://github.com/wjwwood/catkin_demos/tree/master/python_package/foo

This is what I did:

```
% mkdir build

% cd build

% cmake .. -DCMAKE_INSTALL_PREFIX=../../install
-- Using CATKIN_DEVEL_PREFIX: /tmp/catkin_testing/foo/build/devel
-- Using CMAKE_PREFIX_PATH: /Users/william/hydro/install_isolated
-- This workspace overlays: /Users/william/hydro/install_isolated
-- Using default Python package layout
-- Using CATKIN_ENABLE_TESTING: ON
-- Call enable_testing()
-- Using CATKIN_TEST_RESULTS_DIR: /tmp/catkin_testing/foo/build/test_results
-- Found gtest: gtests will be built
-- catkin 0.5.71
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/catkin_testing/foo/build

% make

% make install
Install the project...
-- Install configuration: ""
-- Installing: /tmp/catkin_testing/install/.catkin
-- Installing: /tmp/catkin_testing/install/_setup_util.py
-- Installing: /tmp/catkin_testing/install/env.sh
-- Installing: /tmp/catkin_testing/install/setup.bash
-- Installing: /tmp/catkin_testing/install/setup.sh
-- Installing: /tmp/catkin_testing/install/setup.zsh
-- Installing: /tmp/catkin_testing/install/.rosinstall
-- Installing: /tmp/catkin_testing/install/etc/catkin/profile.d/05.catkin_make.bash
-- Installing: /tmp/catkin_testing/install/etc/catkin/profile.d/05.catkin_make_isolated.bash
+ '[' -n '' ']'
+ cd /tmp/catkin_testing/foo
+ /usr/bin/env PYTHONPATH=/tmp/catkin_testing/install/lib/python2.7/site-packages:/tmp/catkin_testing/foo/build/lib/python2.7/site-packages:/Users/william/hydro/install_isolated/lib/python2.7/site-packages:/usr/local/lib/python2.7/site-packages: CATKIN_BINARY_DIR=/tmp/catkin_testing/foo/build /usr/bin/python /tmp/catkin_testing/foo/setup.py build --build-base /tmp/catkin_testing/foo/build install --prefix=/tmp/catkin_testing/install --install-scripts=/tmp/catkin_testing/install/bin
running build
running build_py
creating /tmp/catkin_testing/foo/build/lib
creating /tmp/catkin_testing/foo/build/lib/foo
copying src/foo/__init__.py -> /tmp/catkin_testing/foo/build/lib/foo
running install
running install_lib
creating /tmp/catkin_testing/install/lib
creating /tmp/catkin_testing/install/lib/python2.7
creating /tmp/catkin_testing/install/lib/python2.7/site-packages
creating /tmp/catkin_testing/install/lib/python2.7/site-packages/foo
copying /tmp/catkin_testing/foo/build/lib/foo/__init__.py -> /tmp/catkin_testing/install/lib/python2.7/site-packages/foo
byte-compiling /tmp/catkin_testing/install/lib/python2.7/site-packages/foo/__init__.py to __init__.pyc
running install_egg_info
Writing /tmp/catkin_testing/install/lib/python2.7/site-packages/foo-0.1.0-py2.7.egg-info
-- Installing: /tmp/catkin_testing/install/lib/pkgconfig/foo.pc
-- Installing: /tmp/catkin_testing/install/share/foo/cmake/fooConfig.cmake
-- Installing: /tmp/catkin_testing/install/share/foo/cmake/fooConfig-version.cmake
-- Installing: /tmp/catkin_testing/install/share/foo/package.xml

% cd ..

% cd ..

% ls
foo install

% cd install

% ls
_setup_util.py  env.sh      etc     lib     setup.bash  setup.sh    setup.zsh   share

% cd lib/python2.7/site-packages/foo

% ls
__init__.py __init__.pyc

% cat __init__.py
# This is a custom file

my_pkg_name = 'foo'
```

**Comment by severin-lemaignan** [comment]
@dirk-thomas : I can reproduce the error with the package 'nao_driver' in the 'nao_robot' stack:
https://github.com/severin-lemaignan/nao_robot

@wjwwood As I mentionned before, I do reproduce the error only on one machine (the Nao virtual machine), not on the other one (my dev laptop). So it's likely that the issue is related to an environment configuration issue. I'm just not sure how to debug that (I'm not too familiar with catkin internals)

**Comment by wjwwood** [comment]
https://github.com/severin-lemaignan/nao_robot/tree/devel

**Comment by dirk-thomas** [comment]
Are you sure that you use the `devel` branch of the repo?

If it is a problem in your workflow it might help if you post the full command sequence starting with a `git clone` of the repo and also include some `ls` / `cat` of your folders (src, build, devel).

**Comment by severin-lemaignan** [comment]
Sorry, yes, the `devel` branch. We've just merged several branches over 3 repos, and I got a bit confused...

Ok, after recompiling the packages from a clean GIT clone, I do not have the issue anymore: as I understand it, for some reason, `make install` would not overwrite a previous (incorrect) `nao_driver/__init__.py` (typically because the file's timestamp was somewhere in the future due to time synchronisation between host and guest machines).

With a fresh clone, it seems to fix the issue.

Sorry for the noise...
